flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1032,13 +1035,16 @@ else if (current == ExecutionState.RUNNING) {
 								producedPartitions,
 								inputGates);
 
-						Thread cancelThread = new Thread(
-								executingThread.getThreadGroup(),
-								canceler,
-								String.format("Canceler for %s (%s).", taskNameWithSubtask, executionId));
-						cancelThread.setDaemon(true);
-						cancelThread.setUncaughtExceptionHandler(FatalExitExceptionHandler.INSTANCE);
-						cancelThread.start();
+						CompletableFuture<Void> future = CompletableFuture.runAsync(canceler,
```

Comment:
Just waiting for cancelation triggering seems to be not enough because task is still running concurrently and it will not solve the original [issue](https://issues.apache.org/jira/browse/FLINK-11631). It looks like we have to introduce a `CompletableFuture<Void> terminationFuture` in `Task` which can be completed when `Task.run()` exits, e.g. in `finally` block. Otherwise, task can continue to use concurrently being freed task executor services.